flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/descriptors/SchemaValidator.java

```diff
@@ -212,12 +212,16 @@ else if (proctimeFound) {
 	@Deprecated
 	public static TableSchema deriveTableSinkSchema(DescriptorProperties properties) {
 		TableSchema.Builder builder = TableSchema.builder();
-
-		TableSchema schema = TableSchemaUtils.getPhysicalSchema(properties.getTableSchema(SCHEMA));
-
-		for (int i = 0; i < schema.getFieldCount(); i++) {
-			TypeInformation t = schema.getFieldTypes()[i];
-			String n = schema.getFieldNames()[i];
+		TableSchema tableSchema = properties.getTableSchema(SCHEMA);
+		for (int i = 0; i < tableSchema.getFieldCount(); i++) {
+			TypeInformation t = tableSchema.getFieldTypes()[i];
+			String n = tableSchema.getFieldNames()[i];
+			Optional<TableColumn> tableColumn = tableSchema.getTableColumn(n);
+			boolean isGeneratedColumn = tableColumn.isPresent() && tableColumn.get().isGenerated();
```

Comment:
Improve the code a bit:

```java
final TableColumn tableColumn = tableSchema.getTableColumns().get(i);
final String fieldName = tableColumn.getName();
final DataType fieldType = tableColumn.getType();
final boolean isGeneratedColumn = tableColumn.isGenerated();
```
We do know it's safe to call `tableSchema.getTableColumns().get(i)`.

Reply:
That's better, thanks for your kindly tips